engine/src/flutter/shell/platform/darwin/macos/framework/Source/FlutterRunLoop.swift

```diff
@@ -0,0 +1,159 @@
+import Foundation
+
+/// Interface for scheduling tasks on the run loop.
+///
+/// The main difference between using `FlutterRunLoop` to schedule tasks compared
+/// to `DispatchQueue.async` or `RunLoop.perform(_:)` is that `FlutterRunLoop`
+/// schedules the task in both common run loop mode and a private run loop mode,
+/// which allows it to run in a mode where it only processes Flutter messages
+/// (`pollFlutterMessagesOnce()`).
+@objc public class FlutterRunLoop: NSObject {
+  private let runLoop: CFRunLoop = CFRunLoopGetCurrent()
+  private var tasks: [Task] = []
+  private let tasksLock = NSLock()
+
+  private var source: CFRunLoopSource!
+  private var timer: CFRunLoopTimer!
+
+  private struct Task {
+    let block: () -> Void
+    let targetTime: CFAbsoluteTime
+  }
+
+  private static let flutterRunLoopMode = CFRunLoopMode("FlutterRunLoopMode" as CFString)
+  private static var _mainRunLoop: FlutterRunLoop?
+
+  private override init() {
+    super.init()
+
+    var sourceContext = CFRunLoopSourceContext(
+      version: 0,
+      info: Unmanaged.passUnretained(self).toOpaque(),
+      retain: nil,
+      release: nil,
+      copyDescription: nil,
+      equal: nil,
+      hash: nil,
+      schedule: nil,
+      cancel: nil,
+      perform: { info in
+        let runner = Unmanaged<FlutterRunLoop>.fromOpaque(info!).takeUnretainedValue()
+        runner.performExpiredTasks()
+      }
+    )
+    guard let createdSource = CFRunLoopSourceCreate(kCFAllocatorDefault, 0, &sourceContext) else {
+      fatalError("Failed to create CFRunLoopSource")
+    }
+    source = createdSource
+    CFRunLoopAddSource(runLoop, source, .commonModes)
+    CFRunLoopAddSource(runLoop, source, Self.flutterRunLoopMode)
+
+    var timerContext = CFRunLoopTimerContext(
+      version: 0,
+      info: Unmanaged.passUnretained(self).toOpaque(),
+      retain: nil,
+      release: nil,
+      copyDescription: nil,
+    )
+    guard let createdTimer = CFRunLoopTimerCreate(
+      kCFAllocatorDefault,
+      CFAbsoluteTime.greatestFiniteMagnitude,
+      CFAbsoluteTime.greatestFiniteMagnitude,
```

Comment:
Personally, I think the right call would be:
```swift
CFAbsoluteTime.distantFuture,
0,
```
but that's a change to the logic and I'd like to keep this code as close to the original as possible.